Hexagonal blog

iluwatar opened the pull request:
This is a pull request for a blog post about Hexagonal Architecture. The code samples have been embedded using [gist-it](http://gist-it.appspot.com/). Looking forward to your comments.

markusmo3 commented:
So i havent read your blogpost from start to finish 100%, as this isnt my style in general, i always scroll back and forth to scan for relevant information in posts and only want to read the good information.

At the very beginning the explanation is good, once read through, without much skipping i got the general idea of the pattern. :+1: 
But when the code explanation starts it gets quite hard with my style of reading, as important parts arent highlighted, or in another way specially presented. The code is just shown in a wall of ~~text~~ code. But if i skip the whole file i miss important information and selectively picking out important parts is hard as mentioned above.

Some code parts are also included that i think of unnecessary, like the LotteryTicket#hashCode #equals, getter methods and imports. I dont think the reader needs those informations right at the moment he/she comes in first contact with the new pattern (well, i didnt at least).

The page-filling amount of code breaks the "flow" for me, its like that for me:
I see a complete source file, think okay, well lets look at it. I understand it, scroll down a bit, see **more** source files, read them a bit, but also get more and more impatient reading through and sifting out important information that i just skipped the last source files and looked up different blog posts about this pattern.

**For me** the blogpost is still valueable, as the code often *does* contain good commenting and explanation. I just think there has to be a better way of representing the informations and comments in different way than just `cat`ing the files.

Dont get discouraged by me! I really :heart: the idea of a blogpost explaining the pattern in detail and i think this is what jdp really needs at this point.

iluwatar commented:
@markusmo3 thanks for the review and valuable comments. I'll try to cut down the amount of code, especially the irrelevant parts. I understand it can be disheartening for the reader to have to read hundreds of lines of code at once.

iluwatar commented:
I have walked through the implementation chapters and reduced the amount of code while adding more explanation.

npathai commented:
Going through it right now. Let u know comments in sometime. Looks promising already :+1:

npathai commented:
![hexagonal](https://cloud.githubusercontent.com/assets/1792515/17669347/fab2957e-632a-11e6-90a0-2c041e07ba45.png)
![lottery system](https://cloud.githubusercontent.com/assets/1792515/17669360/057c6354-632b-11e6-8088-e16ef709ebc0.png)

@iluwatar I have redone the diagrams to make them look pretty. Use em if you wish.

npathai commented:
@iluwatar I have gone through the blog post. It is promising :+1: . 
Here are a few suggestions
 - Images of nut and lottery system are not necessary.
 - You may integrate diagrams that I have attached above.
 - Is `LotterySystem` a port? Or a part of Domain logic itself? I am confused there.
 - We may also attach a detailed diagram of how control flows from ports and adapters.

iluwatar commented:
Thanks a lot for your effort @npathai I will surely integrate the diagrams. I can also try it out without the nut and lottery picture.

`LotterySystem` in the diagram is actually the domain logic or the core. I am not sure if it should be renamed. I'll think about the flow diagram as well.

iluwatar commented:
So I updated the images as follows:
- nut image stays partly because an image is needed on the blog listing page http://java-design-patterns.com/blog/ and partly because I kind of like it :)
- lottery picture was removed
- the diagrams from @npathai were included (they are great!)

npathai commented:
@iluwatar Great, glad you liked the images. If u can wait till this weekend then I will also create introduction image for the blog, so that we will not need nut image. IMHO I dnt feel the image is necessary but again it's ur blog post n ur call 😊 I can give it a shot if u like.

npathai commented:
My doubt regarding lottery system was not regarding the name. Should lottery system be cled a port? And if yes then what will be different adapters of this port?

iluwatar commented:
@npathai I won't publish this weekend so you have time to suggest the new image.

npathai commented:
There you go. Not exactly my idea, saw it being used somewhere, found a creative commons image and edited it to add blogpost title. Hope you like it.

![hexagonal-scaled](https://cloud.githubusercontent.com/assets/1792515/17847065/a0eab726-6869-11e6-9421-8fda8138fdec.png)

npathai commented:
@iluwatar My concern in the present explanation is that the core of the system does not contain any business logic, there are only data structures like `LotteryTicket`, `LotteryNumbers` and `PlayerDetails`. The domain of the system must contain logic that remains the same and through ports and adapters we are able to test the domain and run it in multiple environments without the change leaking into the core.

One of the reason hexagonal architecture is utilized is ease of testing the system without dependency on database or such, but in our system there will be nothing to test but the datastructures. This feels a little weird to me. Shouldn't business logic of how ticket numbers are selected be part of the core domain and not a port? Any insights would be great. Thanks :+1:

iluwatar commented:
@npathai Yes, I tend to agree with you comment that the core should include domain logic. I think I will try to add another class `LotterySystem` that provides interfaces for `LotteryService` and `LotteryAdministration`. Basically the system will work the same but the domain logic will move to the core. Does that sound about right?

I like the new image and will try to use that instead of the controversial nut :grin:

iluwatar commented:
@npathai The work is ready for another review. I've made the changes as described in the previous comment.

@markusmo3 I understood the feature concerning `gist-it` and commit hashes is ready? Could you instruct me how to add those commit hashes?

npathai commented on _posts/2016-09-04-build-maintainable-systems-with-hexagonal-architecture.md:
```diff
@@ -81,7 +95,7 @@ Another secondary port is the notification service. If the player has written hi
 <script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/notifications/LotteryNotificationsImpl.java?slice=26:"></script>
 
-The methods in LotteryNotificationsImpl adapter are simple System.out printers so the implementation is trivial to understand.
+The methods in `LotteryNotificationsImpl` adapter are simple `System.out` printers so the implementation is trivial to understand.
```
A better name for `LotteryNotificationsImpl` would be `ConsoleNotifications`. Also if the interface represents an entity that will send out notifications then the name can be something like `LotteryEventNotifier`, and the console implementation can have name `ConsoleNotifier`.
Also in the code there are `*Impl` classes, I am not fond of names like `*Impl`, rather the implementation should describe how it does, what it does. Plus why do classes like `LotterySystemImpl` hardcode their dependency on `LotteryNotificationsImpl`, `LotteryTicketInMemoryRepository`, `WireTransfersImpl`. Out of these lottery ticket repository and Wire Transfers are dependencies of the system and should be injected using constructor. What if we wanted to test the system using `MongoTicketRepository`?

npathai commented on _posts/2016-09-04-build-maintainable-systems-with-hexagonal-architecture.md:
```diff
@@ -0,0 +1,163 @@
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/notifications/LotteryNotifications.java?slice=26:"></script>
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/notifications/LotteryNotificationsImpl.java?slice=26:"></script>
+
+The methods in `LotteryNotificationsImpl` adapter are simple `System.out` printers so the implementation is trivial to understand.
```
A better name for `LotteryNotificationsImpl` would be `ConsoleNotifications`. Also if the interface represents an entity that will send out notifications then the name can be something like `LotteryEventNotifier`, and the console implementation can have name `ConsoleNotifier` and maybe a `SMSNotifier`.
Also in the code there are `*Impl` classes, I am not fond of names like `*Impl`, rather the implementation should describe how it does, what it does. Plus why do classes like `LotterySystemImpl` hardcode their dependency on `LotteryNotificationsImpl`, `LotteryTicketInMemoryRepository`, `WireTransfersImpl`. Out of these lottery ticket repository and Wire Transfers are dependencies of the system and should be injected using constructor. What if we wanted to test the system using `MongoTicketRepository`?

npathai commented on _posts/2016-09-04-build-maintainable-systems-with-hexagonal-architecture.md:
```diff
@@ -0,0 +1,163 @@
+
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/domain/LotterySystemImpl.java?slice=34:107"></script>
+
+`LotterySystemImpl` has dependencies to lottery ticket database, banking and notifications ports. This implementation hardwires the adapters it uses. In a more sophisticated application we would use dependency injection to determine the implementation classes. The core logic is tested in [LotteryTest](https://github.com/iluwatar/java-design-patterns/blob/master/hexagonal/src/test/java/com/iluwata ...
```
I can understand that the example needs to be kept easier to understand, but as Notifications, Database, and Banking are the ports of the system they ought to be injected into the system and not hardcode the dependencies.

The App (which is the production replica) should use actual implementations such as Email notifier, MongoRepository and such. The AppTest which will wire up the App for testing should use ConsoleNotifier, InMemoryRepository and so on. This will showcase the power of the pattern IMHO.

markusmo3 commented:
@iluwatar Sure.

Have a look at the last comment here: https://github.com/iluwatar/java-design-patterns/issues/486
Its standard jekyll/liquid usage:
```
{% include gistit.html file="visitor/pom.xml" slice="2:4" %}

{% %}: means its a jekyll command
include gistit.html: is the jekyll command to be executed, in our case an include (could also call that an import)
file="visitor/pom.xml" slice="2:4": are include parameters that are given to the gistit.html file
```

Such include parameters for external use are always documented [in the include file itself](https://github.com/iluwatar/java-design-patterns/blob/a4887f9620562f64d349e51de5fa64a6d9f8b6ad/_includes/gistit.html#L1-L16)

Due to jekyll not "really" being a programming language some quirk might sound a bit wierd, but those includes/imports could also be imagined as a method/function, taking parameter and returning html code.

Hope this clears it up a bit.

Further reading: https://jekyllrb.com/docs/templates/#includes

iluwatar commented:
@npathai can you take another look now? I've moved the logic to the core, introduced Mongo based adapters and console user interfaces for players and administrators.

Additionally @npathai , could you update the lottery system picture so that `Notifications` port becomes `Event log`?

npathai commented:
@iluwatar Sure I will do do that. Will do it this weekend.

npathai commented:
![lottery system](https://cloud.githubusercontent.com/assets/1792515/19959326/d4b0217c-a1cb-11e6-9cbe-79373250a257.png)

@iluwatar Here is the updated diagram

npathai commented:
@iluwatar I am done with review. It seems good to go. Great job done on this blog. :+1:

iluwatar commented:
@npathai @markusmo3 thank you very much for the help! The blog post just went live.

iluwatar commented:
@all-contributors please add @markusmo3 for design code ideas
